### DISCUSSION TITLE
Slint + async Rust = ?

### DISCUSSION THREAD

**Discussion by Tmpod** [slint-ui/slint#4377] [Q&A] [answered]
Hello!

In a similar fashion to [this post](https://github.com/KDAB/cxx-qt/discussions/805) I made on CXX-Qt's discussions, I'm interested in knowing how Slint fairs with async Rust.

I've been interested in cross-platform GUIs for a while, particularly with Rust and QtQuick/QML. I discovered Slint at the beginning of last year, but haven't really dived into it yet. However, I'm quite curious, since it's made by Qt folks, is clearly heavily inspired by the modern QtQuick/QML system and has first-party support for Rust.

Async/await is becoming ever more prevalent, with many new API wrappers and the likes being published exclusively as async crates. One particular example I'm interested on is the great [matrix-rust-sdk](https://github.com/matrix-org/matrix-rust-sdk), which is now powering clients such as Element X iOS (Rust+SwiftUI), Element X Android (Rust+Jetpack Compose) and Fractal (Rust+GTK4). I'd like to play around with making a desktop client with Rust and either QtQuick/QML or Slint :)

How does using an async library like the one mentioned above in a Slint UI look like? Is there first-party support? If not, are there plans for such a thing?

On another somewhat related note, are there any major show stoppers in Slint in terms of feature parity with QML, specially with regards to accessibility and portability?

Thank you!

**Comment by hunger** [comment]
I have seen people use tokio with slint, but am no authority on how to do that. AFAICT you set aside one thread for the UI, pretty much like you would in Qt, too. Nothing Slint needs to support too much:-)

I have not really used Qt in a while, so I do not feel qualified to do a comparison -- but here is the Slint side of the picture:

Slint has support for accessibility, but this area could use some more work to make it truly great. In my experience that is the case with most accessibility systems out there though.

The limiting factor for portability of Slint is the rust compiler. In my experience all the interesting things are supported there, but some people will miss their favorite CPU architectures. Slint can be made to work pretty much everywhere the rust compiler can generate code for, it does not even need the full rust standard library. `core` and `alloc` are enough. Out of the box Slint will work on Linux, Mac and Windows, we have a list of supported MCU platforms and Android support is in the works.

Language portability of Slint is very good IMHO: We officially support Rust, C++ and JS as business logic languages and are currently adding python into the mix. A community member has used Slint from C#. We are of course always open for more languages:-)

The Slint language itself has a very strict separation between business logic and UI. I like that a lot. Basically a designer can design to his hearts content and a developer can refactor the code as needed -- as long as the interface between UI and business logic stays untouched, they will not break each others work. It also will allow us to have great tooling (soon;-) and makes support for so many different languages possible.

I am sure compared to Qt we do miss out on features. Most of that is by design: We happily delegate to rust infrastructure (e.g. docs, build tooling, etc.) and other people's crates (== rust libraries). Of course we also have features missing simply because nobody had the time to implement them yet. In my experience most bases are covered, as we tend to focus on those area the most people report issues in;-) And contributions are always welcome, too.

I hope this helps at least a bit.

**Reply by Tmpod** [reply]
Thank you for such a detailed response!

**Comment by ogoffart** [comment, marked as answer]
For async, Slint can run future in the UI thread with the [slint::spawn_local](https://docs.rs/slint/latest/slint/fn.spawn_local.html) function.
With that function you can spawn future  and call async function from a callback handler, for example.

You can also run other async executor (eg, tokio) on another thread. It is fairly easy to call back into the ui thread with slint::invoke_from_event_loop.

**Reply by Tmpod** [reply]
So, with `slint::spawn_local` I can run futures, for example, in a Rust function triggered by a UI click, but what if they depend on Tokio? 
`slint::invoke_from_event_loop` can be used for calling back to the UI thread from an async function, right?

**Reply by ogoffart** [reply]
if the future depends on tokio, then i believe you can run `tokio::task::spawn(...)` with your future from the gui thread as long as you started another thread with the tokio runtime.

**Reply by Tmpod** [reply]
Right, that makes sense, thank you! :3

**Reply by bjorn** [reply]
I've used `tokio::task::spawn` in Raccoin, so I could interface with various async APIs:

https://github.com/bjorn/raccoin/blob/c1198b61c5b29cdc31b0c1599064149b1d485d74/src/main.rs#L1806-L1823

However, I'm not sure whether it is the best way.

I had previously tried to rely only on `slint::spawn_local` (with only a `#[tokio::main]`, otherwise there was a runtime error), but after initially seeming to run an async task fine, it eventually resulted in a locked up application with 100% CPU usage, so I [switched to `tokio::task::spawn`](https://github.com/bjorn/raccoin/commit/679fcfa48f57f168329bb395e57b64aaa9659bdc). This unfortunately meant using `Arc<Mutex<>>` instead of `Rc<RefCell<>>` to wrap my app, since the latter was not `Send`, making it impossible to talk back to my application from the future. I even had to take all `Rc` members out of my app, which is why it now accesses the various models through the UI getters whereas previously I could keep `Rc<VecModel<>>` members around.

It's a bit simpler if your future only needs to talk back to the UI, because `slint::Weak` is `Send`. But unfortunately `slint::Weak` only works with a `ComponentHandle`.

**Reply by ogoffart** [reply]
I wonder if something like this would work:

```rust
my_ui.on_some_callback(move || {
    let my_ui = my_ui_weak.unwrap();
    // ....
    slint::spawn_local(async move {
        // ...
        let foobar = tokio::task::spawn(async move {
             // do the thing that needs to run on a tokio executor
             let foobar = compute_foobar(/*......*/).await;
             foobar
        }).await;
        // now use foobar to set some property
        // ...
        my_ui.set_foobar(foobar);
        });
});
```

We should document this better and maybe have an example.

**Reply by Tmpod** [reply]
I haven't had proper time to dive into this, but looking at those snippets, I'm thinking of making some function or macro to simplify that.

I'll see what I can manage to do. Will post here when I get something decent.

**Reply by bjorn** [reply]
@ogoffart I've tried it, and yes that works as well, thanks!

This way, I can keep using `Rc<VecModel<...>>` since those references no longer needs to be `Send` since it stays within the `slint::spawn_local` future, so I can avoid using `as_any` / `downcast_ref` when getting the model from the UI. And indeed it also avoids the need for `slint::invoke_from_event_loop`, since the future is only executed there in the first place.

When documenting this, maybe there could be a word about why directly using `slint::spawn_local` to run the future does not work when using Tokio? Rust async is still a bit of a mystery to me, especially the part where you can have different runtimes and the compiler will not catch any errors caused by doing something the runtime does not expect you to.

**Reply by ogoffart** [reply]
Trying to improve the docs in https://github.com/slint-ui/slint/pull/4407

**Reply by ti-integra** [reply]
this help too much. this specific not work for me but work when use runtime over task (I am not using async main)
```rust
let rt = tokio::runtime::Runtime::new().unwrap();
let footbar = rt
         .spawn(async move { pegar_dados(&input).await })
         .await
         .unwrap(); // this work
// let footbar = tokio::task::spawn(async move { pegar_dados(&input).await })
// .await
// .unwrap(); // this for me crash with "there is no reactor running, must be called from the context of a Tokio 1.x runtime"
```

**Reply by ogoffart** [reply]
To be honest i'm not so familiar with Tokio either.
Maybe you will have to call rt.enter() or something.
And you must do that in the bottom of your stack (typically in main.rs, before starting the slit event loop.

In the GUI thread you must call slint::spawn_local if you want to get the result.

The recently added doc is there:

https://slint.dev/snapshots/master/docs/rust/slint/fn.spawn_local#compatibility-with-tokio-and-other-runtimes

**Reply by Tmpod** [reply]
> I've tried it, and yes that works as well, thanks!

I saw that, and was giving it a shot, but it still requires the Tokio spawned future to be `Send` (which isn't possible with `Rc`). How did it work for you?

**Reply by bjorn** [reply]
It worked for me, because I did not need to pass a reference to my app in the Tokio future. It sufficed to await the Tokio task, and to only refer back to my app and its models from the async block passed to `slint::spawn_local`. I think this would be commonly the case.

Unfortunately my patch wasn't very polished yet so I can't yet link to my code for reference.

**Reply by Tmpod** [reply]
I see. I took a better look at your code and I think I understand why you didn't have issues: you create a client on each interaction, **inside** tokio's spawn block. ([example](https://github.com/bjorn/raccoin/blob/master/src/main.rs#L1824))

In my case, I want to have a long lived client object (actually managed by an external crate) which provides async methods I need to call inside a tokio runtime.

**Reply by Tmpod** [reply]
I didn't end up doing much yesterday, but I ended up cooking this, which is working (compiles and runs):

```rust
let ui = AppWindow::new()?;
let app = Arc::new(App { client: make_client() });

ui.on_foo({
    let ui_weak = ui.as_weak();
    let app_ref = app.clone();
    move || {
        let ui = ui_weak.unwrap();
        let app = app_ref.clone();
        let _ = slint::spawn_local(async move {
            let foo = tokio::spawn(async move { app.client.foo().await }).await.unwrap();
            ui.set_foo(foo.into());
        });
    }
});
```
Where `App` is just a struct with some APIs wrapper client.

However, this feels a bit clunky and with too many `Weak`/`Arc` clones. Can this be simplified?  
I've also made some macros to heavily simplify this in terms of syntax, but I was looking to simplifying the logic itself.

**Reply by Tmpod** [reply]
I'm having issues with another thing, though: changing UI from within tokio tasks (through `slint::invoke_from_event_loop`).
Specifically, the API client I'm using requires their event handlers to be both `Send` and `Sync`, and it seems `Weak` isn't the latter, so I can't get a weak ref for the UI component inside the closure. How can this be solved?

Edit: Sorry if I'm hijacking this thread way too much! Let me know if I should create a new one.

**Comment by maxenko** [comment]
Just to add to above example, as it took a bit to wrap my mind around.

```rust
    runtime.clone().on_state_change.lock().await.subscribe({
        let ui_handle = ui_handle.clone();
        move |state| {
            let ui_handle = ui_handle.clone();
            slint::invoke_from_event_loop(move || {
                let ui = ui_handle.clone().unwrap();
               // do stuff with ui...
            }).unwrap();
        }

    }).await;
```

Here I queue up a lambda that will be invoked from a different thread to be run on main UI thread. This is more or less identical to how WPF dispatches calls from other threads on main ui thread (through a message queue). **Make sure you are not sending ui handles to methods with Sync traits (as Slint is not multi-threaded).**

In short, there are no issues spawning threads from main UI thread, or having parallel threads make calls to slint ui.

**Comment by Tmpod** [comment]
Nice, thank you for the extra snippet!
For the project I'm working on, because of all the repetitiveness and
verbosity of this approach, I made some macros that greatly simplify
callbacks and so on. I'm still not 100% sold on them, but they're here:
https://git.sr.ht/~tmpod/eigen/tree/main/item/src/utils.rs

**Reply by Tmpod** [reply]
(apparently replies via e-mail don't get added to the existing comment thread :/)

**Comment by Areopagitics** [comment]
Here is a working example updated from [here](https://github.com/DaMilyutin/sixtyfps-rust-template-worker-threads).

I would be interested to hear if this would be the proper "slint" way to do async.

The UI:
```
import {ListView, Button, VerticalBox, Slider } from "std-widgets.slint";


export struct ListItemData {
    id: int,
    progress: float,
}

export component AppWindow inherits Window {
    preferred-width: 400px;
    preferred-height: 600px;
    
    in-out property<int> counter: 42;
    in-out property<float> latency <=> slider.value;
    in-out property<[ListItemData]> task-data-model;

    callback request-increase-value();

    HorizontalLayout {
        VerticalBox {
            width: parent.width/2;
            Text { text: "Counter: \{counter}"; }
            Button {
                text: "Increase value";
                clicked => {
                    request-increase-value();
                }
            }
            Text{ text: "Latency: \{slider.value}"; }
            slider := Slider {
                height: 20px;
            }
        }

        ListView {
            for task[i] in task-data-model:
                Rectangle {
                    border-color: #030C8FF1;
                    border-width: 1px;
                    VerticalLayout {
                        Text { text: "[\{i}] task: \{task.id} => progress: \{task.progress}"; }                        
                        Rectangle { 
                            width: task.progress/100*parent.width;
                            height: 3px;
                            background: task.progress > 99.9? #00FF00: #FF0000;   
                        }
                    }
                }
        }

    }
}
```

The Main Rust:
```

slint::include_modules!();

use slint::{Model, ModelRc, VecModel};
use std::rc::Rc;
use std::sync::{Arc, Mutex};
use std::thread;
use std::time;

mod id_map_data;
use id_map_data::nextId;
use id_map_data::IdMapData;

fn main() {
    let ui = AppWindow::new().unwrap();

    let ui_handle = ui.as_weak();

    let list_data: Rc<VecModel<ListItemData>> = Rc::new(VecModel::<ListItemData>::default());
    let task_data = Arc::new(Mutex::new(IdMapData::default()));

    ui.set_task_data_model(ModelRc::new(list_data.clone()));

    ui.on_request_increase_value({
        let task_data = task_data.clone();
        move || {
            let ui = ui_handle.unwrap();
            let id = nextId() as i32;
            let model_handle = ui.get_task_data_model();
            let data = model_handle
                .as_any()
                .downcast_ref::<VecModel<ListItemData>>()
                .unwrap();
            task_data.lock().as_deref_mut().unwrap().push(data, id);
            let latency = ui.get_latency();
            let period = time::Duration::from_millis((latency * 10.) as u64);

            let ui_handle = ui.as_weak();

            let task_data = task_data.clone();

            std::thread::spawn(move || {
                for i in 0..101 {
                    thread::sleep(period);
                    let task_data = task_data.clone();
                    let ui_handle = ui_handle.clone();
                    slint::invoke_from_event_loop(move || {
                        let ui = ui_handle.unwrap();
                        let model_handle = ui.get_task_data_model();
                        let data = model_handle
                            .as_any()
                            .downcast_ref::<VecModel<ListItemData>>()
                            .unwrap();
                        task_data
                            .lock()
                            .as_deref_mut()
                            .unwrap()
                            .set_progress(data, id, i as f32)
                    }).unwrap();
                }
                thread::sleep(time::Duration::from_secs(1));
                slint::invoke_from_event_loop(move || {
                    let ui = ui_handle.unwrap();
                    let model_handle = ui.get_task_data_model();
                    let data = model_handle
                        .as_any()
                        .downcast_ref::<VecModel<ListItemData>>()
                        .unwrap();
                    task_data
                        .lock()
                        .as_deref_mut()
                        .unwrap()
                        .remove_by_id(data, id);
                    ui.set_counter(ui.get_counter() + 1);
                }).unwrap();
            });
        }
    });

    ui.run().unwrap();
}

```

AND id_map_data.rs

```
use super::*;
use slint::Model;
use std::{collections::HashMap, vec::Vec};

pub fn nextId() -> i32 {
    unsafe {
        static mut n: i32 = 0 as i32;
        const period: i32 = (1 << 15) as i32;
        n = (n + 1i32) % period;
        n
    }
}

pub struct IdMapData {
    pub id2row: HashMap<i32, usize>,
    pub row2id: Vec<i32>,
}

impl IdMapData {
    pub fn default() -> IdMapData {
        IdMapData {
            id2row: HashMap::<i32, usize>::default(),
            row2id: Vec::<i32>::default(),
        }
    }

    pub fn push(&mut self, data: &slint::VecModel<ListItemData>, id: i32) {
        self.row2id.push(id);
        self.id2row.insert(id, self.row2id.len() - 1);
        data.push(ListItemData {
            id: id,
            progress: 0f32,
        })
    }

    pub fn set_progress(&self, data: &slint::VecModel<ListItemData>, id: i32, progress: f32) {
        let row = *self.id2row.get(&id).unwrap();
        data.set_row_data(
            row,
            ListItemData {
                id: id,
                progress: progress,
            },
        );
    }

    pub fn remove_by_id(&mut self, data: &slint::VecModel<ListItemData>, id: i32) {
        let row = *self.id2row.get(&id).unwrap();
        self.remove_by_row(data, row);
    }

    pub fn remove_by_row(&mut self, data: &slint::VecModel<ListItemData>, row: usize) {
        assert!(0 <= row && row < self.row2id.len());
        let mut r = row + 1;
        while r < self.row2id.len() {
            let id = self.row2id.get(r);
            *self.id2row.get_mut(id.unwrap()).unwrap() -= 1;
            r += 1;
        }
        self.id2row.remove(&self.row2id.get(row).unwrap());
        self.row2id.remove(row);
        data.remove(row);
    }

    pub fn quit(&mut self, data: &slint::VecModel<ListItemData>) {
        let r = self.row2id.len();
        while r != 0 {
            self.remove_by_row(data, r - 1);
        }
    }
}

```
